### PR TITLE
Proving lemmas doc

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: docs/2020-02-05-Proving-Lemmas.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+```
+We would prove them by induction on BS. The set of constructors/generators, is,
+of course, `nilBytes` and `:`.
+
+Let’s start with the first claim. We will start by defining a macro:
+```
+P(i, BS) :=  lengthBytes(BS, i) ==Int i +Int lengthBytes(BS)
+```
+
+Next, we will attempt to prove `P(i, nilBytes)` and
```
Comment (traiansf) [inline docs/2020-02-05-Proving-Lemmas.md]: Maybe justify better how induction is used here, by clearly stating the property proved by induction and the goals derived by induction from it.

File: docs/2020-02-05-Proving-Lemmas.md [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+==============
+
+Background
+----------
+
+In order to prove rewrite claims, we usually need lemmas describing
+the function symbols we use. As an example, we may have a lemma saying that the
+length of a string is not negative, i.e.
+```
+rule length(S) > 0 => true
```
Comment (ttuegel) [inline docs/2020-02-05-Proving-Lemmas.md]:
```suggestion
rule length(S) >= 0 => true
```

File: docs/2020-02-05-Proving-Lemmas.md [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+```
+
+The lemmas that I have seen are, usually, given without proofs.
+
+There is an effort by daejunpark and malturki to prove some properties
+for Casper, simulating Coq-style proofs:
+https://github.com/runtimeverification/beacon-chain-verification/tree/static-model-coq-specs-dependencies/casper/k
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply docs/2020-02-05-Proving-Lemmas.md]: Hm, removing this feels kind of bad... Oh well.

File: docs/2020-02-05-Proving-Lemmas.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+```
+
+Next, we will attempt to prove `forall i . P(i, nilBytes)` and
+`P(i, B:BS) if (forall j . P(j, BS))`.
+To do that, we will use the following rewrite rule: `start(X) => end(X)`
+and the following hand-written claims:
+```
+start(P(i, nilBytes)) => end(true)
+start(P(i, B:BS)) => end(true)
+    if P(i + 1, BS) and P(1, BS)
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply docs/2020-02-05-Proving-Lemmas.md]: Why? I think need `P(1, BS)` in the proof below. I added an explanation as a code comment, does it make more sense now?

File: docs/2020-02-05-Proving-Lemmas.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+```
+We will prove it by structural induction on `BS`. To do that, let us define
+a macro:
+```
+P(i, BS) :=  lengthBytes(BS, i) ==Int i +Int lengthBytes(BS)
+```
+
+Next, we will attempt to prove `forall i . P(i, nilBytes)` and
+`P(i, B:BS) if (forall j . P(j, BS))`.
+To do that, we will use the following rewrite rule: `start(X) => end(X)`
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply docs/2020-02-05-Proving-Lemmas.md]: Better now?

File: docs/2020-02-05-Proving-Lemmas.md (hunk 2 above)
Comment (virgil-serbanuta) [reply docs/2020-02-05-Proving-Lemmas.md]: Thanks!

**Comment by virgil-serbanuta** [comment]
Thanks for the comments, I think I fixed them all, I'll probably merge tomorrow unless you have new suggestions/comments.

**Pull request merged**